### PR TITLE
postgres-protocol: add pg numeric type

### PR DESCRIPTION
**Pull request by fredr**
A first stab at adding support for the Numeric type as well as a to/from implementation for string as that is kind of the only std type that can hold a numeric.

Would appreciate if you had the time to review this. Happy to fix any feedback. I'm thinking any ToSql and FromSql implementations can be added as followups to this one.

The protocol is quite strait forward, the string conversion is more finicky, but added a bunch of test to hopefully catch all corner cases. Let me know if you think there is any tests that should be added for those.

### CODE REVIEW

File: postgres-protocol/src/types/numeric.rs [hunk 1]
```diff
@@ -0,0 +1,1583 @@
...
+    if leading_zero_count > 0 {
+        if leading_zero_count == digits.len() {
+            // All digits are zero
+            *weight = 0;
+            digits.clear();
+        } else {
+            // Remove leading zeros
+            digits.drain(0..leading_zero_count);
+            *weight -= leading_zero_count as i16;
+        }
```
Comment (paolobarbolini) [inline postgres-protocol/src/types/numeric.rs]: Is this micro-optimization worth it over always going through `drain`?

File: postgres-protocol/src/types/numeric.rs [hunk 2]
```diff
@@ -0,0 +1,1583 @@
...
+fn split_e(s: &[u8]) -> (&[u8], Option<&[u8]>) {
+    let mut s = s.splitn(2, |&b| b == b'e' || b == b'E');
+    let first = s.next().unwrap();
+    let second = s.next();
+    (first, second)
+}
+
+fn split_decimal(s: &[u8]) -> (&[u8], Option<&[u8]>) {
+    let mut s = s.splitn(2, |&b| b == b'.');
+    let first = s.next().unwrap();
```
Comment (paolobarbolini) [inline postgres-protocol/src/types/numeric.rs]: Quickly scrolling through the code it's non-obvious that this panic is unreachable.

File: postgres-protocol/src/types/numeric.rs [hunk 3]
```diff
@@ -0,0 +1,1583 @@
...
+        if e.is_empty() {
+            return Err("empty scientific notation exponent".into());
+        }
+
+        for &b in e {
+            if !b.is_ascii_digit() {
+                return Err("scientific notation string contain non-digit character".into());
+            }
+            // Prevent integer overflow in exponent
+            exp = exp.saturating_mul(10).saturating_add((b - b'0') as i32);
```
Comment (paolobarbolini) [inline postgres-protocol/src/types/numeric.rs]: Can this really overflow if we're checking for `exp` not to get too big in the next line?

File: postgres-protocol/src/types/numeric.rs [hunk 4]
```diff
@@ -0,0 +1,1583 @@
...
+            let numeric = Numeric::from_str(&prop_val).expect("parse numeric");
+            let str_val = numeric.to_string();
+            assert_eq!(prop_val, str_val, "proprty test value {value}");
+        }
+        #[test]
+        fn test_arbitrary_i64_from_string_and_back(value in any::<i64>()) {
+            let prop_val = value.to_string();
+            let numeric = Numeric::from_str(&prop_val).expect("parse numeric");
+            let str_val = numeric.to_string();
+            assert_eq!(prop_val, str_val, "proprty test value {value}");
```
Comment (paolobarbolini) [inline postgres-protocol/src/types/numeric.rs]:
```suggestion
            assert_eq!(prop_val, str_val, "property test value {value}");
```

File: postgres-protocol/src/types/numeric.rs [hunk 5]
```diff
@@ -0,0 +1,1583 @@
...
+            NumericSign::Positive => 0x0000,
+            NumericSign::Negative => 0x4000,
+            NumericSign::NaN => 0xC000,
+            NumericSign::PositiveInfinity => 0xD000,
+            NumericSign::NegativeInfinity => 0xF000,
+        }
+    }
+}
+
+/// A Posgres numeric
```
Comment (paolobarbolini) [inline postgres-protocol/src/types/numeric.rs]: Copies the description from `PgLsn`.

```suggestion
/// Postgres `NUMERIC` type.
```

**Review by paolobarbolini — COMMENTED** [review]
I've left a couple of comments, including one on how you see this get used.

The final version should also have tests in `tokio-postgres/tests/test/types/` and commits squashed.

File: postgres-protocol/src/types/numeric.rs [hunk 6]
```diff
@@ -0,0 +1,1583 @@
...
+        return Ok(Some(Numeric::nan()));
+    }
+    if s.eq_ignore_ascii_case("Infinity") || s.eq_ignore_ascii_case("Inf") {
+        return Ok(Some(Numeric::infinity()));
+    }
+    if s.eq_ignore_ascii_case("-Infinity") || s.eq_ignore_ascii_case("-Inf") {
+        return Ok(Some(Numeric::negative_infinity()));
+    }
+    Ok(None)
+}
```
[the remark this answers is not in the document]
Comment (fredr) [reply postgres-protocol/src/types/numeric.rs]: good catch!

File: postgres-protocol/src/types/numeric.rs (hunk 3 above)
Comment (fredr) [reply postgres-protocol/src/types/numeric.rs]: very true, good catch!

**Comment by fredr** [comment]
Thanks for reviewing this! I'll wait with squashing until we are happy with the PR. Regarding tests in `tokio-postgres`, from what I can tell, those are more tests of types that implement `ToSql` and `FromSql`, so not sure exactly what I should add for tests for this pr? I assume when we start adding `bigdecimal` and other types, those will be tested there, but not the protocol type? or should I add something there that uses `FromStr` and `Display`?
